### PR TITLE
feat(web): add function editor with lsp

### PR DESCRIPTION
**Pull request by newfish-cmyk**

### CODE REVIEW

File: web/src/pages/functionTemplate/Mods/MonacoEditor/index.tsx [hunk 1]
```diff
@@ -199,7 +127,65 @@ const MonacoEditor = (props: {
...
+              "editorIndentGuide.background": "#eee",
+            },
+          });
+
+          monaco.editor.defineTheme("MonacoEditorThemeDark", {
+            base: "vs-dark",
+            inherit: true,
+            rules: [
+              {
+                foreground: "65737e",
```
Comment (0fatal) [inline web/src/pages/functionTemplate/Mods/MonacoEditor/index.tsx:165]: #65737e

File: web/src/pages/app/mods/StatusBar/LSPBar/index.tsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
+import { useMemo } from "react";
+import { useTranslation } from "react-i18next";
+import { Spinner, Tooltip } from "@chakra-ui/react";
+
+import { createUrl, createWebSocketAndStartClient } from "@/components/Editor/LanguageClient";
+
+import useFunctionStore from "@/pages/app/functions/store";
+import useGlobalStore from "@/pages/globalStore";
+export default function LSPBar() {
+  const hostname = "scm3dt.100.66.76.85.nip.io";
```
Comment (0fatal) [inline web/src/pages/app/mods/StatusBar/LSPBar/index.tsx:10]: hostname

File: web/src/pages/app/functions/mods/FunctionPanel/index.tsx [hunk 3]
```diff
@@ -312,6 +314,12 @@ export default function FunctionList() {
                         const res = await deleteFunctionMutation.mutateAsync(item);
                         if (!res.error) {
                           showSuccess(t("DeleteSuccess"));
+
```
Comment (0fatal) [inline web/src/pages/app/functions/mods/FunctionPanel/index.tsx:317]: delete

File: web/src/components/TextButton/index.tsx [hunk 4]
```diff
@@ -14,8 +14,7 @@ const TextButton = function (
     <Button
       variant={"link"}
       size="xs"
-      colorScheme={"primary"}
-      className={className}
+      className={className + " !text-[#219BF4]"}
```
Comment (0fatal) [inline web/src/components/TextButton/index.tsx:17]: clsx

File: web/src/components/Editor/userWorker.ts [hunk 5]
```diff
@@ -25,4 +24,4 @@ self.MonacoEnvironment = {
   },
 };
 
-monaco.languages.typescript.typescriptDefaults.setEagerModelSync(true);
+// monaco.languages.typescript.typescriptDefaults.setEagerModelSync(true);
```
Comment (0fatal) [inline web/src/components/Editor/userWorker.ts:27]: remove

File: web/src/components/Editor/userWorker.ts [hunk 6]
```diff
@@ -1,11 +1,10 @@
-import * as monaco from "monaco-editor";
+// import * as monaco from 'monaco-editor';
```
Comment (0fatal) [inline web/src/components/Editor/userWorker.ts:1]: remove

File: web/src/components/Editor/theme/index.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
+import { registerExtension } from "vscode/extensions";
+
+var manifest = {
+  name: "theme-append",
+  displayName: "Default Themes",
+  description: "The default Visual Studio light and dark themes",
+  categories: ["Themes"],
+  version: "1.0.0",
+  publisher: "vscode",
```
Comment (0fatal) [inline web/src/components/Editor/theme/index.js:9]: publisher

File: web/src/components/Editor/TextModel.tsx [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+  input: any,
+) {
+  let model = this.findModel(editor, input.resource);
+  if (!model) {
+    return null;
+  }
+
+  editor.setModel(model);
+
+  const functionName = model.uri.path.match(/\/functions\/([^\.]+)\.ts/)[1];
```
Comment (0fatal) [inline web/src/components/Editor/TextModel.tsx:31]: lint

File: web/src/components/Editor/LanguageClient.ts [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+};
+export const performInit = async (vscodeApiInit: boolean) => {
+  if (vscodeApiInit === true) {
+    await initServices({
+      userServices: {
+        ...getThemeServiceOverride(),
+        ...getTextmateServiceOverride(),
+        ...getConfigurationServiceOverride(Uri.file(RUNTIMES_PATH)),
+      },
+      // debugLogging: true,
```
Comment (0fatal) [inline web/src/components/Editor/LanguageClient.ts:83]: remove

File: web/src/components/Editor/LanguageClient.ts [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+      reader,
+      writer,
+    });
+    await languageClient.start();
+    reader.onClose(() => languageClient?.stop());
+  };
+  return webSocket;
+};
+
+export type ExampleJsonEditor = {
```
Comment (0fatal) [inline web/src/components/Editor/LanguageClient.ts:69]: remove

File: web/src/components/Editor/LanguageClient.ts [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+  webSocket.onopen = async () => {
+    webSocket.send(token);
+    const socket = toSocket(webSocket);
+    const reader = new WebSocketMessageReader(socket);
+    const writer = new WebSocketMessageWriter(socket);
+    const languageClient = createLanguageClient({
+      reader,
+      writer,
+    });
+    await languageClient.start();
```
Comment (0fatal) [inline web/src/components/Editor/LanguageClient.ts:63]: remove await

File: web/src/components/Editor/LanguageClient.ts [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+      },
+    },
+  });
+};
+
+export const createUrl = (
+  hostname: string,
+  port: number,
+  path: string,
+  secure?: boolean,
```
Comment (0fatal) [inline web/src/components/Editor/LanguageClient.ts:44]: secure?: boolean = window.isSecureContext

File: web/src/components/Editor/LanguageClient.ts [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+        error: () => ({ action: ErrorAction.Continue }),
+        closed: () => ({ action: CloseAction.DoNotRestart }),
+      },
+      workspaceFolder: {
+        uri: Uri.file(RUNTIMES_PATH),
+        name: "Laf Workspace",
+        index: 0,
+      },
+    },
+    // create a language client connection from the JSON RPC connection on demand
```
Comment (0fatal) [inline web/src/components/Editor/LanguageClient.ts:31]: remove

File: web/src/components/Editor/LanguageClient.ts [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+import { toSocket, WebSocketMessageReader, WebSocketMessageWriter } from "vscode-ws-jsonrpc";
+
+import { RUNTIMES_PATH } from "@/constants";
+
+import "@codingame/monaco-vscode-typescript-basics-default-extension";
+import "./TextModel";
+
+export const createLanguageClient = (transports: MessageTransports): MonacoLanguageClient => {
+  return new MonacoLanguageClient({
+    name: "Sample Language Client",
```
Comment (0fatal) [inline web/src/components/Editor/LanguageClient.ts:18]: Laf Language Client

File: web/src/components/Editor/FunctionEditor.tsx [hunk 15]
```diff
@@ -130,14 +55,20 @@ function FunctionEditor(props: {
   const editorRef = useRef<monaco.editor.IStandaloneCodeEditor | null>();
   const subscriptionRef = useRef<monaco.IDisposable | undefined>(undefined);
   const monacoEl = useRef(null);
-
+  const hostname = "scm3dt.100.66.76.85.nip.io";
```
Comment (0fatal) [inline web/src/components/Editor/FunctionEditor.tsx:58]: hostname

File: web/src/components/Editor/FunctionEditor.tsx [hunk 16]
```diff
@@ -146,67 +77,139 @@ function FunctionEditor(props: {
...
       }
-    }
-  }, [path, value, isFetchButtonClicked, setIsFetchButtonClicked]);
+    };
+
+    window.addEventListener("unhandledrejection", listener);
+    return () => {
+      window.removeEventListener("unhandledrejection", listener);
+    };
+  });
```
Comment (0fatal) [inline web/src/components/Editor/FunctionEditor.tsx:118]: , []

File: web/src/components/Editor/FunctionEditor.tsx [hunk 17]
```diff
@@ -146,67 +77,139 @@ function FunctionEditor(props: {
...
 
   useEffect(() => {
-    if (monacoEl && editorRef.current) {
-      editorRef.current.updateOptions({
-        fontSize: fontSize,
-        theme: colorMode === COLOR_MODE.dark ? "lafEditorThemeDark" : "lafEditorTheme",
-      });
+    if (monacoEl && !editorRef.current) {
+      const start = async () => {
```
Comment (0fatal) [inline web/src/components/Editor/FunctionEditor.tsx:122]: remove start wrap

File: web/src/components/Editor/FunctionEditor.tsx [hunk 18]
```diff
@@ -146,67 +77,139 @@ function FunctionEditor(props: {
...
+          fontFamily: "Fira Code",
+          fontWeight: "450",
+          scrollBeyondLastLine: false,
+        });
+      };
+      start();
     }
-  }, [colorMode, fontSize]);
+
+    allFunctionList.forEach(async (item: any) => {
```
Comment (0fatal) [inline web/src/components/Editor/FunctionEditor.tsx:147]: remove async, change any

File: web/src/components/Editor/FunctionEditor.tsx [hunk 19]
```diff
@@ -146,67 +77,139 @@ function FunctionEditor(props: {
...
+          monaco.editor.createModel(
+            functionCache.getCache(item._id, item.source?.code),
+            "typescript",
+            uri,
+          );
+        }
+      }
+    });
+
+    functionList.forEach(async (item: any) => {
```
Comment (0fatal) [inline web/src/components/Editor/FunctionEditor.tsx:162]: remove async, change any

**Review by 0fatal — CHANGES_REQUESTED** [review]

File: web/src/components/Editor/FunctionEditor.tsx [hunk 20]
```diff
@@ -9,13 +9,12 @@ import { Position } from "vscode/vscode/src/vs/editor/common/core/position";
 
 import { COLOR_MODE, Pages, RUNTIMES_PATH } from "@/constants";
 
-import "./userWorker";
-
 // import "./theme"
```
Comment (0fatal) [inline web/src/components/Editor/FunctionEditor.tsx:12]: remove

File: web/src/components/Editor/FunctionEditor.tsx [hunk 21]
```diff
@@ -55,17 +54,15 @@ function FunctionEditor(props: {
...
     (state) => state,
-  );
-  const functionCache = useFunctionCache();
-  const [functionList, setFunctionList] = useState(allFunctionList);
-
+    );
+    const functionCache = useFunctionCache();
+    const [functionList, setFunctionList] = useState(allFunctionList);
+    const hostname = globalStore.currentApp.origin;
+    const url = useMemo(() => createUrl(hostname, 80, "/_/lsp"), [hostname]);
```
Comment (0fatal) [inline web/src/components/Editor/FunctionEditor.tsx:64]: remove 80

File: web/src/components/Editor/TextModel.tsx [hunk 22]
```diff
@@ -28,23 +28,25 @@ StandaloneCodeEditorService.prototype.doOpenEditor = function (
 
   editor.setModel(model);
 
-  const functionName = model.uri.path.match(/\/functions\/([^\.]+)\.ts/)[1];
+  const functionName = model.uri.path.match(/\/functions\/([^.]+)\.ts/)[1];
```
Comment (0fatal) [inline web/src/components/Editor/TextModel.tsx:31]: \.

**Review by 0fatal — CHANGES_REQUESTED** [review]

**Review by 0fatal — APPROVED** [review]

**Review by maslow — APPROVED** [review]

**Pull request merged**